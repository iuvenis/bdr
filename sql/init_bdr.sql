\c postgres

SELECT bdr.bdr_is_active_in_db();

SELECT bdr.bdr_group_create(
	local_node_name := 'node-pg',
	node_external_dsn := 'dbname=postgres user=postgres',
	replication_sets := ARRAY['default', 'important', 'for-node-1']
	);

SELECT bdr.bdr_is_active_in_db();

SELECT bdr.bdr_node_join_wait_for_ready();

SELECT bdr.bdr_is_active_in_db();

\c regression

SELECT bdr.bdr_is_active_in_db();

SELECT bdr.bdr_group_join(
	local_node_name := 'node-regression',
	node_external_dsn := 'dbname=regression user=postgres',
	join_using_dsn := 'dbname=postgres user=postgres',
	node_local_dsn := 'dbname=regression user=postgres',
	replication_sets := ARRAY['default', 'important', 'for-node-2', 'for-node-2-insert', 'for-node-2-update', 'for-node-2-delete']
	);

SELECT bdr.bdr_is_active_in_db();

SELECT * FROM  bdr.global_lock_info();

SELECT bdr.bdr_node_join_wait_for_ready();

SELECT bdr.bdr_is_active_in_db();

SELECT owner_replorigin, (owner_sysid, owner_timeline, owner_dboid) = bdr.bdr_get_local_nodeid(), lock_mode, lock_state, owner_local_pid = pg_backend_pid() AS owner_pid_is_me, lockcount, npeers, npeers_confirmed, npeers_declined, npeers_replayed, replay_upto IS NOT NULL AS has_replay_upto FROM bdr.global_lock_info();

-- Make sure we see two slots and two active connections
SELECT plugin, slot_type, database, active FROM pg_replication_slots;
SELECT count(*) FROM pg_stat_replication;

\c postgres
SELECT conn_dsn, conn_replication_sets FROM bdr.bdr_connections ORDER BY conn_dsn;
SELECT node_status, node_local_dsn, node_init_from_dsn FROM bdr.bdr_nodes ORDER BY node_local_dsn;

SELECT 1 FROM bdr.pg_replication_slots WHERE restart_lsn <= confirmed_flush_lsn;

\c regression
SELECT conn_dsn, conn_replication_sets FROM bdr.bdr_connections ORDER BY conn_dsn;
SELECT node_status, node_local_dsn, node_init_from_dsn FROM bdr.bdr_nodes ORDER BY node_local_dsn;

SELECT 1 FROM bdr.pg_replication_slots WHERE restart_lsn <= confirmed_flush_lsn;

SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE VIEW public.ddl_info AS
SELECT
	owner_replorigin,
	(owner_sysid, owner_timeline, owner_dboid) = bdr.bdr_get_local_nodeid() AS is_my_node,
	lock_mode,
	lock_state,
	owner_local_pid IS NOT NULL AS owner_pid_set,
	owner_local_pid = pg_backend_pid() AS owner_pid_is_me,
	lock_state = 'acquire_acquired' AND owner_local_pid = pg_backend_pid() AS fully_owned_by_me,
	lockcount,
	npeers,
	npeers_confirmed,
	npeers_declined,
	npeers_replayed,
	replay_upto IS NOT NULL AS has_replay_upto
FROM bdr.global_lock_info();
$DDL$);

BEGIN;

SELECT * FROM ddl_info;

SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE OR REPLACE FUNCTION public.bdr_regress_variables(
    OUT readdb1 text,
    OUT readdb2 text,
    OUT writedb1 text,
    OUT writedb2 text
    ) RETURNS record LANGUAGE SQL AS $f$
SELECT
    current_setting('bdrtest.readdb1'),
    current_setting('bdrtest.readdb2'),
    current_setting('bdrtest.writedb1'),
    current_setting('bdrtest.writedb2')
$f$;
$DDL$);

SELECT * FROM ddl_info;

COMMIT;

SELECT * FROM ddl_info;

-- Run the upgrade function, even though we started with 2.0, so we exercise it
-- and so we know it won't break things when run on a 2.0 cluster.
SELECT bdr.upgrade_to_200();
