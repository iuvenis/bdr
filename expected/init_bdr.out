\c postgres
SELECT bdr.bdr_is_active_in_db();
 bdr_is_active_in_db 
---------------------
 f
(1 row)

SELECT bdr.bdr_group_create(
	local_node_name := 'node-pg',
	node_external_dsn := 'dbname=postgres user=postgres',
	replication_sets := ARRAY['default', 'important', 'for-node-1']
	);
WARNING:  Secondary unique constraint(s) exist on replicated table(s)
DETAIL:  Table public.some_local_tbl has secondary unique constraint some_local_tbl_key_key. This may cause unhandled replication conflicts.
HINT:  Drop the secondary unique constraint(s), change the table(s) to UNLOGGED if they don't need to be replicated, or exclude the table(s) from the active replication set(s).
 bdr_group_create 
------------------
 
(1 row)

SELECT bdr.bdr_is_active_in_db();
 bdr_is_active_in_db 
---------------------
 t
(1 row)

SELECT bdr.bdr_node_join_wait_for_ready();
 bdr_node_join_wait_for_ready 
------------------------------
 
(1 row)

SELECT bdr.bdr_is_active_in_db();
 bdr_is_active_in_db 
---------------------
 t
(1 row)

\c regression
SELECT bdr.bdr_is_active_in_db();
 bdr_is_active_in_db 
---------------------
 f
(1 row)

SELECT bdr.bdr_group_join(
	local_node_name := 'node-regression',
	node_external_dsn := 'dbname=regression user=postgres',
	join_using_dsn := 'dbname=postgres user=postgres',
	node_local_dsn := 'dbname=regression user=postgres',
	replication_sets := ARRAY['default', 'important', 'for-node-2', 'for-node-2-insert', 'for-node-2-update', 'for-node-2-delete']
	);
 bdr_group_join 
----------------
 
(1 row)

SELECT bdr.bdr_is_active_in_db();
 bdr_is_active_in_db 
---------------------
 t
(1 row)

SELECT * FROM  bdr.global_lock_info();
ERROR:  database with oid=16385 is not configured for bdr or bdr is still starting up
SELECT bdr.bdr_node_join_wait_for_ready();
 bdr_node_join_wait_for_ready 
------------------------------
 
(1 row)

SELECT bdr.bdr_is_active_in_db();
 bdr_is_active_in_db 
---------------------
 t
(1 row)

SELECT owner_replorigin, (owner_sysid, owner_timeline, owner_dboid) = bdr.bdr_get_local_nodeid(), lock_mode, lock_state, owner_local_pid = pg_backend_pid() AS owner_pid_is_me, lockcount, npeers, npeers_confirmed, npeers_declined, npeers_replayed, replay_upto IS NOT NULL AS has_replay_upto FROM bdr.global_lock_info();
 owner_replorigin | ?column? | lock_mode | lock_state | owner_pid_is_me | lockcount | npeers | npeers_confirmed | npeers_declined | npeers_replayed | has_replay_upto 
------------------+----------+-----------+------------+-----------------+-----------+--------+------------------+-----------------+-----------------+-----------------
                  | f        |           | nolock     | f               |         0 |      1 |                0 |               0 |               0 | f
(1 row)

-- Make sure we see two slots and two active connections
SELECT plugin, slot_type, database, active FROM pg_replication_slots;
 plugin | slot_type |  database  | active 
--------+-----------+------------+--------
 bdr    | logical   | postgres   | t
 bdr    | logical   | regression | t
(2 rows)

SELECT count(*) FROM pg_stat_replication;
 count 
-------
     2
(1 row)

\c postgres
SELECT conn_dsn, conn_replication_sets FROM bdr.bdr_connections ORDER BY conn_dsn;
            conn_dsn             |                                conn_replication_sets                                 
---------------------------------+--------------------------------------------------------------------------------------
 dbname=postgres user=postgres   | {default,important,for-node-1}
 dbname=regression user=postgres | {default,important,for-node-2,for-node-2-insert,for-node-2-update,for-node-2-delete}
(2 rows)

SELECT node_status, node_local_dsn, node_init_from_dsn FROM bdr.bdr_nodes ORDER BY node_local_dsn;
 node_status |         node_local_dsn          |      node_init_from_dsn       
-------------+---------------------------------+-------------------------------
 r           | dbname=postgres user=postgres   | 
 r           | dbname=regression user=postgres | dbname=postgres user=postgres
(2 rows)

SELECT 1 FROM bdr.pg_replication_slots WHERE restart_lsn <= confirmed_flush_lsn;
 ?column? 
----------
        1
        1
(2 rows)

\c regression
SELECT conn_dsn, conn_replication_sets FROM bdr.bdr_connections ORDER BY conn_dsn;
            conn_dsn             |                                conn_replication_sets                                 
---------------------------------+--------------------------------------------------------------------------------------
 dbname=postgres user=postgres   | {default,important,for-node-1}
 dbname=regression user=postgres | {default,important,for-node-2,for-node-2-insert,for-node-2-update,for-node-2-delete}
(2 rows)

SELECT node_status, node_local_dsn, node_init_from_dsn FROM bdr.bdr_nodes ORDER BY node_local_dsn;
 node_status |         node_local_dsn          |      node_init_from_dsn       
-------------+---------------------------------+-------------------------------
 r           | dbname=postgres user=postgres   | 
 r           | dbname=regression user=postgres | dbname=postgres user=postgres
(2 rows)

SELECT 1 FROM bdr.pg_replication_slots WHERE restart_lsn <= confirmed_flush_lsn;
 ?column? 
----------
        1
        1
(2 rows)

SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE VIEW public.ddl_info AS
SELECT
	owner_replorigin,
	(owner_sysid, owner_timeline, owner_dboid) = bdr.bdr_get_local_nodeid() AS is_my_node,
	lock_mode,
	lock_state,
	owner_local_pid IS NOT NULL AS owner_pid_set,
	owner_local_pid = pg_backend_pid() AS owner_pid_is_me,
	lock_state = 'acquire_acquired' AND owner_local_pid = pg_backend_pid() AS fully_owned_by_me,
	lockcount,
	npeers,
	npeers_confirmed,
	npeers_declined,
	npeers_replayed,
	replay_upto IS NOT NULL AS has_replay_upto
FROM bdr.global_lock_info();
$DDL$);
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

BEGIN;
SELECT * FROM ddl_info;
 owner_replorigin | is_my_node | lock_mode | lock_state | owner_pid_set | owner_pid_is_me | fully_owned_by_me | lockcount | npeers | npeers_confirmed | npeers_declined | npeers_replayed | has_replay_upto 
------------------+------------+-----------+------------+---------------+-----------------+-------------------+-----------+--------+------------------+-----------------+-----------------+-----------------
                  | f          |           | nolock     | t             | f               | f                 |         0 |      1 |                0 |               0 |               0 | f
(1 row)

SELECT bdr.bdr_replicate_ddl_command($DDL$
CREATE OR REPLACE FUNCTION public.bdr_regress_variables(
    OUT readdb1 text,
    OUT readdb2 text,
    OUT writedb1 text,
    OUT writedb2 text
    ) RETURNS record LANGUAGE SQL AS $f$
SELECT
    current_setting('bdrtest.readdb1'),
    current_setting('bdrtest.readdb2'),
    current_setting('bdrtest.writedb1'),
    current_setting('bdrtest.writedb2')
$f$;
$DDL$);
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

SELECT * FROM ddl_info;
 owner_replorigin | is_my_node | lock_mode |    lock_state    | owner_pid_set | owner_pid_is_me | fully_owned_by_me | lockcount | npeers | npeers_confirmed | npeers_declined | npeers_replayed | has_replay_upto 
------------------+------------+-----------+------------------+---------------+-----------------+-------------------+-----------+--------+------------------+-----------------+-----------------+-----------------
                0 | t          | ddl_lock  | acquire_acquired | f             |                 |                   |         1 |      1 |                0 |               0 |               0 | f
(1 row)

COMMIT;
SELECT * FROM ddl_info;
 owner_replorigin | is_my_node | lock_mode | lock_state | owner_pid_set | owner_pid_is_me | fully_owned_by_me | lockcount | npeers | npeers_confirmed | npeers_declined | npeers_replayed | has_replay_upto 
------------------+------------+-----------+------------+---------------+-----------------+-------------------+-----------+--------+------------------+-----------------+-----------------+-----------------
                  | f          |           | nolock     | t             | f               | f                 |         0 |      1 |                0 |               0 |               0 | f
(1 row)

-- Run the upgrade function, even though we started with 2.0, so we exercise it
-- and so we know it won't break things when run on a 2.0 cluster.
SELECT bdr.upgrade_to_200();
 upgrade_to_200 
----------------
 
(1 row)

